### PR TITLE
Implement jump threading MIR opt

### PR DESCRIPTION
**Pull request by cjgillot**
This pass is an attempt to generalize `ConstGoto` and `SeparateConstSwitch` passes into a more complete jump threading pass.

This pass is rather heavy, as it performs a truncated backwards DFS on MIR starting from each `SwitchInt` terminator. This backwards DFS remains very limited, as it only walks through `Goto` terminators.

It is build to support constants and discriminants, and a propagating through a very limited set of operations.

The pass successfully manages to disentangle the `Some(x?)` use case and the DFA use case. It still needs a few tests before being ready.

### CODE REVIEW

File: compiler/rustc_mir_transform/src/jump_threading.rs [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+
+pub struct JumpThreading;
+
+const MAX_BACKTRACK: usize = 5;
+const MAX_COST: usize = 100;
+
+impl<'tcx> MirPass<'tcx> for JumpThreading {
+    fn is_enabled(&self, sess: &rustc_session::Session) -> bool {
+        sess.mir_opt_level() >= 4
+    }
```
Comment (saethlin) [inline compiler/rustc_mir_transform/src/jump_threading.rs:51]: Is there any particular reason we need limits on here? Is the runtime potentially infinite, or just O(big) if we have none?

In general, I think it would be neat if we could have _some_ mechanism to turn off all such limits. For DataflowConstProp I suggested that pass be enabled with limits at opt-level 3, and enabled with no limits at level 4. Of course I'm not aware of any agreement on what opt levels are for so maybe I'm the odd one here.

**Comment by nikic** [comment]
Just to check, does this implementation prevent the formation of irreducible control flow?

**Comment by cjgillot** [comment]
@nikic I don't know. I'm not familiar enough with the literature, and I haven't put anything specific to prevent them, so I guess they are not prevented. Do you have pointers on prior work so I can verify/modify the pass properly?

**Comment by JakobDegen** [comment]
> https://en.wikipedia.org/wiki/Control-flow_graph#Reducibility

See the [Wikipedia entry](https://en.wikipedia.org/wiki/Control-flow_graph#Reducibility) for the definition of reducibility. I'll review this in a bit though and can make a suggestion then as to how to preserve it

File: compiler/rustc_mir_transform/src/jump_threading.rs [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+
+    fn insert_ref(&mut self, place: PlaceIndex, conds: &'a [Condition]) {
+        self.state.insert(place, conds);
+    }
+
+    fn take(&mut self, place: PlaceIndex, map: &Map) -> Option<&'a [Condition]> {
+        let current_condition = self.state.remove(&place);
+        map.preorder_invoke(place, &mut |place| {
+            self.state.remove(&place);
+        });
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/jump_threading.rs:174]: This does not invalidate enough places for discriminants to be sound. Ideally, a write to an enum field should invalidate all the other variants and the discriminant of the enum recursively, and reciprocally.
This means that `discriminant((_1 as variant#1).0) = ?` invalidates `_1 as other variants` and `((_1 as variant#1).0 as other variants)`.

**Comment by eholk** [comment]
I'm probably not enough of a MIR expert to review this... Re-rolling.

@bors r? compiler

File: compiler/rustc_mir_transform/src/jump_threading.rs [hunk 3]
```diff
@@ -0,0 +1,638 @@
...
+                if let Some(lhs) = self.map.find(lhs_place.as_ref()) {
+                    match rhs {
+                        Rvalue::Use(operand) => self.process_operand(bb, lhs, operand, state)?,
+                        // Transfer the conditions on the copy rhs.
+                        Rvalue::CopyForDeref(rhs) => {
+                            self.process_operand(bb, lhs, &Operand::Copy(*rhs), state)?
+                        }
+                        Rvalue::Discriminant(rhs) => {
+                            let rhs = self.map.find_discr(rhs.as_ref())?;
+                            state.insert_place_idx(rhs, lhs, self.map);
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/jump_threading.rs:375]: This should perform a union between `ConditionSet`s, not just a replacement.
(That's not incorrect, just suboptimal.)

**Comment by apiraino** [comment]
Visiting this PR, unsure its status so tentatively switching to waiting on author. Feel free to request a review explicitely if some parts needs a review (f.e. [this comment](https://github.com/rust-lang/rust/pull/107009#issuecomment-1398061663)?) with `@rustbot ready`, thanks.

@rustbot author

**Comment by wesleywiser** [comment]
Appologies, I don't think I'm going to be able to review this anytime soon so I will reroll. 

r? rust-lang/mir-opt

**Comment by pnkfelix** [comment]
discussed in [today's triage meeting](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-04-06/near/347440185)

We're letting the review assignment stay where it is for now, but if this hasn't been reviewed by the time May rolls around, I'll plan to pick it up then.

**Comment by JakobDegen** [comment]
Again, sorry this has taken me so long to get to. You had asked me to comment on the soundness of this, so I'll start by doing that. The strategy used by this PR looks basically sound to me. I definitely applaud the reuse of the const prop tooling - much of the difficulty of implementing this correctly is basically pushed into there, and having to maintain that logic only once is much better. I have not looked at all of the little details just yet, but I expect that if I do find soundness issues, they'll be bugs in the implementation as opposed to anything fundamental.

That having been said, I'm a little bit concerned that this pass is somewhat over-engineered. Looking at the test cases for a second: The `duplicate_chain` and `multiple_match` cases are actually const-prop. If they are not covered by constprop already, that would be the ideal place to implement them. There's of course nothing wrong with this opt covering them too, but I don't think that should be the primary goal.

For all the other tests, the blocks which we thread over have: 1) a switchint terminator, 2) possibly a `Discriminant` call, and 3) possibly some storage statements. A version of this pass that fully restricts itself to such blocks sounds like it would be significantly simpler. It would of course also be less powerful. Can you add some e2e tests the demonstrate the need for the additional generalization that this is doing at the moment?

**Comment by cjgillot** [comment]
> That having been said, I'm a little bit concerned that this pass is somewhat over-engineered. Looking at the test cases for a second: The `duplicate_chain` and `multiple_match` cases are actually const-prop. If they are not covered by constprop already, that would be the ideal place to implement them. There's of course nothing wrong with this opt covering them too, but I don't think that should be the primary goal.
> 

I agree with you, those particular cases should be handled by dataflow-aware const-prop. Most tests here are from bugs in earlier implementation, not really demonstrating the motivation.

> For all the other tests, the blocks which we thread over have: 1) a switchint terminator, 2) possibly a `Discriminant` call, and 3) possibly some storage statements. A version of this pass that fully restricts itself to such blocks sounds like it would be significantly simpler. It would of course also be less powerful. Can you add some e2e tests the demonstrate the need for the additional generalization that this is doing at the moment?
> 

I added a commit which handles the other terminators.

For the effect on existing pre-codegen test: https://github.com/cjgillot/rust/commit/jump-threading-default
I'll look for more interesting motivational cases.

**Comment by cjgillot** [comment]
r? compiler

**Comment by apiraino** [comment]
rerolling assignment since @TaKO8Ki is currently out of review rotation

r? compiler

**Comment by wesleywiser** [comment]
I'm very sorry, I haven't had time in the last two weeks to look at this and I don't think I will in the next two weeks either. I'm going to reroll so at least I'm not continuing to hold up review efforts. 

r? rust-lang/compiler

**Comment by pnkfelix** [comment]
@JakobDegen [wrote](https://github.com/rust-lang/rust/pull/107009#issuecomment-1528898418):

> For all the other tests, the blocks which we thread over have: 1) a switchint terminator, 2) possibly a `Discriminant` call, and 3) possibly some storage statements. A version of this pass that fully restricts itself to such blocks sounds like it would be significantly simpler. It would of course also be less powerful. Can you add some e2e tests the demonstrate the need for the additional generalization that this is doing at the moment?

Jakob, do you still need more tests beyond the ones that I think @cjgillot added when they added more support for more terminators in https://github.com/rust-lang/rust/pull/107009/commits/048f9dc6c754273072f8163ad5593aa887433dd9 ?

**Comment by pnkfelix** [comment]
(I'll take this rather than let @compiler-errors get even more stuff on their plate.)

r? @pnkfelix

File: compiler/rustc_mir_dataflow/src/value_analysis.rs [hunk 4]
```diff
@@ -482,27 +497,40 @@ impl<V: Clone + HasTop + HasBottom> State<V> {
...
 
-    pub fn flood_discr(&mut self, place: PlaceRef<'_>, map: &Map) {
+    pub fn flood_discr(&mut self, place: PlaceRef<'_>, map: &Map)
+    where
+        V: HasTop,
+    {
         self.flood_discr_with(place, map, V::TOP)
     }
 
+    pub fn flood_with_extra(
```
Comment (pnkfelix) [inline compiler/rustc_mir_dataflow/src/value_analysis.rs:521]: I'm not thrilled about pub method named `flood_with_extra` without any documentation. I cannot tell what the "extra" denotes when I look at the signature, at least not without looking at every other `flood` method.

After review, I infer that this is meant to be the most general form of `flood[_with]`, and every variation on `flood` should call into it (potentially indirectly) right?

(To be fair, I don't know if we have a standard convention for what to name such "core methods". )

But I'd either 1. call the pub one "flood_with_tail_elem" (since that's all that's being added from the viewpoint of the public user of the API), or 2. add some docs to the `flood_with_extra`, if only to make it clear that it may get new things added (or changed) as the internals of the value_analysis module changes.

File: compiler/rustc_mir_dataflow/src/value_analysis.rs [hunk 5]
```diff
@@ -553,44 +581,87 @@ impl<V: Clone + HasTop + HasBottom> State<V> {
...
     }
 
     /// Retrieve the value stored for a place, or ⊤ if it is not tracked.
-    pub fn get(&self, place: PlaceRef<'_>, map: &Map) -> V {
+    pub fn try_get(&self, place: PlaceRef<'_>, map: &Map) -> Option<V> {
+        let place = map.find(place)?;
+        self.try_get_idx(place, map)
+    }
+
+    /// Retrieve the value stored for a place, or ⊤ if it is not tracked.
```
Comment (pnkfelix) [inline compiler/rustc_mir_dataflow/src/value_analysis.rs:611]: Nit: this documentation block for `try_get_discr` is a cut-and-paste of that for `try_get`. I assume it should instead say "Retrieve the value stored for the discriminant of the given place, ..."

File: compiler/rustc_mir_dataflow/src/value_analysis.rs [hunk 6]
```diff
@@ -553,44 +581,87 @@ impl<V: Clone + HasTop + HasBottom> State<V> {
...
+    where
+        V: HasTop,
+    {
         self.flood_discr(target, map);
         if let Some(target) = map.find_discr(target) {
             self.insert_idx(target, result, map);
         }
     }
 
     /// Retrieve the value stored for a place, or ⊤ if it is not tracked.
```
Comment (pnkfelix) [inline compiler/rustc_mir_dataflow/src/value_analysis.rs:605]: Nit: this documentation does not distinguish when `None` is returned.

I assume it meant to say "`None` if it is not tracked" instead of "⊤ if it is not tracked" ?

File: compiler/rustc_mir_dataflow/src/value_analysis.rs [hunk 7]
```diff
@@ -553,44 +581,87 @@ impl<V: Clone + HasTop + HasBottom> State<V> {
...
+        self.try_get_idx(place, map)
+    }
+
+    /// Retrieve the value stored for a place, or ⊤ if it is not tracked.
+    pub fn try_get_discr(&self, place: PlaceRef<'_>, map: &Map) -> Option<V> {
+        let place = map.find_discr(place)?;
+        self.try_get_idx(place, map)
+    }
+
+    /// Retrieve the value stored for a place index, or ⊤ if it is not tracked.
```
Comment (pnkfelix) [inline compiler/rustc_mir_dataflow/src/value_analysis.rs:617]: Same as feedback for `try_get` above: the case where this returns `None` should be distinguished. (And I continue to assume it is a typo and that ⊤ is only returned if ⊤ *is* the value being tracked for the given index.)

Comment (pnkfelix) [inline compiler/rustc_mir_dataflow/src/value_analysis.rs:617]: Same as feedback for `try_get` above: the case where this returns `None` should be distinguished. (And I continue to assume it is a typo and that ⊤ is only returned if ⊤ *is* the value being tracked for the given index.)

File: compiler/rustc_mir_dataflow/src/value_analysis.rs [hunk 8]
```diff
@@ -553,44 +581,87 @@ impl<V: Clone + HasTop + HasBottom> State<V> {
...
+            }
+        }
+    }
+
+    /// Retrieve the value stored for a place, or ⊤ if it is not tracked.
+    pub fn get(&self, place: PlaceRef<'_>, map: &Map) -> V
+    where
+        V: HasBottom + HasTop,
+    {
         map.find(place).map(|place| self.get_idx(place, map)).unwrap_or(V::TOP)
```
Comment (pnkfelix) [inline compiler/rustc_mir_dataflow/src/value_analysis.rs:635]: should this now just be `self.try_get(place, map).unwrap_or(V::TOP)` ?

Comment (cjgillot) [reply compiler/rustc_mir_dataflow/src/value_analysis.rs]: Not exactly. There are cases where it returns BOTTOM. I tried to clarify the code flow.

File: compiler/rustc_mir_transform/src/cost_checker.rs [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+const INSTR_COST: usize = 5;
+const CALL_PENALTY: usize = 25;
+const LANDINGPAD_PENALTY: usize = 50;
+const RESUME_PENALTY: usize = 45;
+
+/// Verify that the callee body is compatible with the caller.
+///
+/// This visitor mostly computes the inlining cost,
+/// but also needs to verify that types match because of normalization failure.
```
Comment (WaffleLapkin) [inline compiler/rustc_mir_transform/src/cost_checker.rs:13]: is this outdated? can't see where the "verify that types match" happens here...

File: compiler/rustc_mir_transform/src/jump_threading.rs [hunk 10]
```diff
@@ -0,0 +1,727 @@
...
+    /// Extract the mutated place from a statement.
+    #[instrument(level = "trace", skip(self), ret)]
+    fn mutated_statement(
+        &self,
+        stmt: &Statement<'tcx>,
+    ) -> Option<(Place<'tcx>, Option<TrackElem>)> {
+        match stmt.kind {
+            StatementKind::Assign(box (place, _))
+            | StatementKind::Intrinsic(box NonDivergingIntrinsic::Assume(
+                Operand::Copy(place) | Operand::Move(place),
```
Comment (pnkfelix) [inline compiler/rustc_mir_transform/src/jump_threading.rs:284]: Hmm.

Question: I can understand how a `Move` operand would be interpreted as a mutation of the place being moved, but why is a `Copy` of a place being interpreted as a mutation of the place being copied?

File: compiler/rustc_mir_transform/src/jump_threading.rs [hunk 11]
```diff
@@ -0,0 +1,727 @@
...
+            ))
+            | StatementKind::Deinit(box place) => Some((place, None)),
+            StatementKind::SetDiscriminant { box place, variant_index: _ } => {
+                Some((place, Some(TrackElem::Discriminant)))
+            }
+            StatementKind::StorageLive(local) | StatementKind::StorageDead(local) => {
+                Some((Place::from(local), None))
+            }
+            StatementKind::Retag(..)
+            | StatementKind::Intrinsic(..)
```
Comment (pnkfelix) [inline compiler/rustc_mir_transform/src/jump_threading.rs:294]: Is it okay for `StatementKind::Intrinsic(box NonDivergingIntrinsic::CopyNonOverlapping(..))` to end up matching this arm? I would have thought that would treat the `dst` of the `CopyNonOverlapping` as mutated.

File: compiler/rustc_mir_transform/src/jump_threading.rs (hunk 10 above)
Comment (pnkfelix) [reply compiler/rustc_mir_transform/src/jump_threading.rs:284]: Although now that I think about it, I'm no longer sure that the place being moved should be treated as a mutation for the purposes of what `mutated_place` is doing...

if I understand correctly, the purpose of `mutated_place` is to cover cases like this during the backward walk over a block:

```
 _1 = 5 // Whatever happens here, it won't change the result of a `SwitchInt` later on
 _1 = 6 // because the SwitchInt will always be based on this assignment to `_1` 
```

So what you care about are mutations that definitively assign a fresh value to a place, not mutations that might or might not overwrite previous values.

Do I have that right?

----

In any case, it would be useful for you to spell out in its doc comment whether `mutated_place` is meant to produce an overapproximation ("these are all the places that *might* be mutated") or an underapproximation ("this is a list of places all of which are *guaranteed* to be initialized by this statement). Having that in my mental model would have made it easier for me to understand the rest of this code.

File: compiler/rustc_mir_transform/src/jump_threading.rs (hunk 11 above)
Comment (pnkfelix) [reply compiler/rustc_mir_transform/src/jump_threading.rs]: (But then again, if `mutated_place` result is meant allow underapproximation (i.e. "this set of places, a subset of those mutated, are *guaranteed* to be initialized by this statement"), then it should always be sound to return `None`.)

File: compiler/rustc_mir_transform/src/jump_threading.rs (hunk 10 above)
Comment (pnkfelix) [reply compiler/rustc_mir_transform/src/jump_threading.rs]: > So what you care about are mutations that definitively assign a fresh value to a place, not mutations that might or might not overwrite previous values.
> 
> Do I have that right?

(It seems I did *not* have that right. 😄 )

File: compiler/rustc_mir_transform/src/jump_threading.rs [hunk 12]
```diff
@@ -0,0 +1,752 @@
...
+        rhs: &Operand<'tcx>,
+        state: &mut State<ConditionSet<'a>>,
+    ) -> Option<!> {
+        let register_opportunity = |c: Condition| {
+            debug!(?bb, ?c.target, "register");
+            self.opportunities.push(ThreadingOpportunity { chain: vec![bb], target: c.target })
+        };
+
+        match rhs {
+            // If we expect `lhs ?= A`, we have an opportunity if we assume `constant == A`.
```
Comment (pnkfelix) [inline compiler/rustc_mir_transform/src/jump_threading.rs:339]: I see the "lhs ?= rhs" notation in many places in this file.

Is that meant to denote an input that is fed into the set of replacement conditions that you describe in the comment block at the start of this file? I.e. one of the cases where assigning `rhs` to `lhs` turns the `SwitchInt` into `Goto { target }`?

I think you should consider adding an explicit definition of the `lhs ?= rhs` notation, since it is going to be ambiguous to future readers. It can go either in the comment block at the start of the file, or in a comment block above this impl block. (The choice of where it should go is up to you based on wheher the notation will be generally useful to people trying to understand this transformation, or if it is only useful for understanding the guts of the implementation.)

Comment (pnkfelix) [reply compiler/rustc_mir_transform/src/jump_threading.rs:339]: Or maybe `lhs ?= A` means "we expect to do a SwitchInt on lhs, seeing if it is equal to A"

That might make more sense given how I see the notation used.

(But that just goes to show the ambiguity of the notation. Which isn't a problem, as long as you define it somewhere.)

File: compiler/rustc_mir_transform/src/jump_threading.rs [hunk 13]
```diff
@@ -17,6 +17,9 @@
 //!   if assigning anything different from `value` to `place` turns the `SwitchInt`
 //!   into `Goto { target }`.
 //!
+//! In this file, we denote as `place ?= value` the existence of a replacement condition
```
[the remark this answers is not in the document]
Comment (pnkfelix) [reply compiler/rustc_mir_transform/src/jump_threading.rs:20]: Actually, it clearly must mean the former, because the notation makes no mention of the target, and therefore the actual effect is not specified and thus the polarity is irrelevant.

File: compiler/rustc_mir_transform/src/jump_threading.rs [hunk 14]
```diff
@@ -17,6 +17,9 @@
...
+//! on `place` with given `value`.
+//!
```
Comment (pnkfelix) [inline compiler/rustc_mir_transform/src/jump_threading.rs:22]:
```suggestion
//! on `place` with given `value`, irrespective of the polarity and target of that
//! replacement condition.
//!
```

File: compiler/rustc_mir_transform/src/jump_threading.rs [hunk 15]
```diff
@@ -0,0 +1,755 @@
...
+            // We are inside `target_bb`. Since we have a single predecessor, we know we passed
+            // through the `SwitchInt` before arriving here. Therefore, we know that
+            // `discr == value`. If one condition can be fulfilled by `discr == value`,
+            // that's an opportunity.
+            for c in conditions.iter_matches(value) {
+                debug!(?target_bb, ?c.target, "register");
+                self.opportunities.push(ThreadingOpportunity { chain: vec![], target: c.target });
+            }
+        } else if target_bb == targets.otherwise() {
+            let (value, _, _) = targets.as_static_if()?;
```
Comment (pnkfelix) [inline compiler/rustc_mir_transform/src/jump_threading.rs:574]: this *might* be clearer if you did `let (value, _, else_) = targets.as_static_if()?;`and then `debug_assert_eq!target_bb, else_)`. At least, I had to double check the body of `otherwise()` and `as_static_if()` to double check that the two methods were synchronized in this manner.

But that also might just be adding unnecessary noise for people deeply versed in these MIR APIs.

File: compiler/rustc_mir_transform/src/jump_threading.rs [hunk 16]
```diff
@@ -0,0 +1,755 @@
...
+            let value = ScalarInt::try_from_uint(value, discr_layout.size)?;
+
+            // Likewise, we know that `discr != value`. That's a must weaker information,
```
Comment (pnkfelix) [inline compiler/rustc_mir_transform/src/jump_threading.rs:577]: What is "That's a must weaker information" supposed to mean here?

Is it supposed to say "That's a much weaker piece of information" ? If so, it is a little bit weird, since the conclusion "we can only match the exact same condition" does not sound you've had to adapt to weaker information.

I nonetheless think i understand the code as a whole; in this case, the branch to the else block will only take place if the tested value *doesn't* match the value you extracted via the `as_static_if` call above.

File: compiler/rustc_mir_transform/src/jump_threading.rs [hunk 17]
```diff
@@ -0,0 +1,755 @@
...
+    #[instrument(level = "trace", skip(self, body))]
+    fn apply_once(&mut self, index: usize, body: &mut Body<'_>) {
+        debug!(?self.predecessors);
+        debug!(?self.involving_tos);
+
+        // Check that `predecessors` satisfies its invariant.
+        debug_assert_eq!(self.predecessors, predecessor_count(body));
+
+        // Remove the TO from the vector to allow modifying the other ones later.
+        let op = &mut self.opportunities[index];
```
Comment (pnkfelix) [inline compiler/rustc_mir_transform/src/jump_threading.rs:630]: Is the comment here left over from a previous version?

I don't see any sign of you "removing" the `op` from `self.opportunities`.

Comment (pnkfelix) [reply compiler/rustc_mir_transform/src/jump_threading.rs:630]: Yes okay that *must* be the case (I.e. the take` is the important thing), because I now see the mutable borrows of `self.opportunities[to_index]` circa line 686 below.

File: compiler/rustc_mir_transform/src/jump_threading.rs [hunk 18]
```diff
@@ -0,0 +1,755 @@
...
+        debug!(?op);
+        let op_chain = std::mem::take(&mut op.chain);
+        let op_target = op.target;
+        debug_assert_eq!(op_chain.len(), op_chain.iter().collect::<FxHashSet<_>>().len());
+
+        let Some((current, chain)) = op_chain.split_first() else { return };
+        let basic_blocks = body.basic_blocks.as_mut();
+
+        // Invariant: we never change the meaning of the program.
```
Comment (pnkfelix) [inline compiler/rustc_mir_transform/src/jump_threading.rs:639]: I'm not sure this Invariant is helpful as its currently phrased. Never changing the meaning of a program certainly is a *goal*, but the comment here doesn't help me understand how you are accomplishing that in this code.

usually when I see invariants in a program, it is describing some condition that is being repeatedly established (e.g. a loop invariant where it may be violated in the middle of a loop but always holds on entry, on re-entry, and on exit)

File: compiler/rustc_mir_transform/src/jump_threading.rs [hunk 19]
```diff
@@ -0,0 +1,755 @@
...
+        let mut current = *current;
+        for &succ in chain {
+            debug!(?current, ?succ);
+
+            // `succ` must be a successor of `current`. If it is not, this means this TO is not
+            // satisfiable, so we bail out.
+            if basic_blocks[current].terminator().successors().find(|s| *s == succ).is_none() {
+                debug!("impossible");
```
Comment (pnkfelix) [inline compiler/rustc_mir_transform/src/jump_threading.rs:650]: Just so I'm clear: This debug output is indicating that it is "impossible" for the Threading Opportunity to be satisfied.

It is *not* indicating that that this code path in the compiler itself is expected to be impossible i.e. unreachable, right?

**Comment by pnkfelix** [comment]
Okay!

I'm pretty much ready to r+ this, I think, based on my survey of the code and the changes @cjgillot has made in response

One final question that I'm not clear on: @cjgillot , the analysis has written has some bits that are sensitive to the number of predecessor arm; e.g. the local performance optimization for a single successor. I haven't thought through it fully, but you might have an immediate answer: If there were no threshold limits (i.e. the analysis were permitted to just run without bounds), would the transformation be idempotent?

To put it another way: do the transformation's changes to the control graph open up new opportunities for the transformation to be applied (which would imply the transformation is *not* idempotent).

My current hypothesis is "an unbounded version of this transformation would be idempotent", based on my cursory understanding of how the analysis and its thoroughness. But I figured you could confirm this.

**Comment by pnkfelix** [comment]
In fact I am ready to r+ this.

@bors r+

File: compiler/rustc_mir_transform/src/jump_threading.rs (hunk 19 above)
Comment (cjgillot) [reply compiler/rustc_mir_transform/src/jump_threading.rs:650]: This pass is actually reachable. I expanded the comment.

File: compiler/rustc_mir_transform/src/jump_threading.rs [hunk 20]
```diff
@@ -0,0 +1,759 @@
...
+/// But if the CFG is already irreducible, there is no point in trying much harder.
+/// is already irreducible.
+fn loop_headers(body: &Body<'_>) -> BitSet<BasicBlock> {
+    let mut loop_headers = BitSet::new_empty(body.basic_blocks.len());
+    let dominators = body.basic_blocks.dominators();
+    // Only visit reachable blocks.
+    for (bb, bbdata) in traversal::preorder(body) {
+        for succ in bbdata.terminator().successors() {
+            if dominators.dominates(succ, bb) {
+                loop_headers.insert(succ);
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/jump_threading.rs:754]: This was wrong. Added a correction commit.

**Comment by cjgillot** [comment]
There is one case where idempotence is broken: when recursing through switchInt terminators. We only recurse into a switchInt if it is the current block's single predecessor. For instance:
```mir
bb0: a = 1; goto bb1;
bb1: switchInt(a) -> [0: bb2, 1: bb2, otherwise: bb5];
bb2: switchInt(a) -> [0: bb3, otherwise: bb4];
```

The first application will transform bb1 into a `goto bb2`, and do nothing to bb2. The second application will transform bb2 to a `goto bb4`..

File: compiler/rustc_mir_transform/src/jump_threading.rs (hunk 20 above)
Comment (pnkfelix) [reply compiler/rustc_mir_transform/src/jump_threading.rs:754]: So... the effect of this change is two-fold (when succ dominates bb): 1. it avoids adding `bb` to the loop headers, and 2. it adds `succ` to the loop headers.

I see essentially only one effect on the tests though (where that one effect is duplicated in two twos) -- and I *think* the effect demonstrated by the tests is due to the addition of adding `succ` to the loop headers (since that cancels out a threading opportunity, which I believe is what is happening in the panic-abort and panic-unwind test)

So, does this imply that there is some missing test coverage, i.e. a test that demonstrates how removing `bb` from loop headers *enables* some jump threading to proceed?

Comment (pnkfelix) [reply compiler/rustc_mir_transform/src/jump_threading.rs:754]: (I'm still going to r+ the PR as is, because the missing test, I think, only represents a failure to demonstrate a newly present optimization opportunity, which while nice to have in tests, is not as important as tests that are catching potential soundness issues in the transformation.)

**Comment by pnkfelix** [comment]
> We only recurse into a switchInt if it is the current block's single predecessor.

Hmm, I think I see what you are saying. The emphasis in the above quote is on "current", since the recursion doesn't itself change the notion of the "current block" (which I think makes sense, given what the transformation is doing), and so you won't recur more than once.

Okay. That seems fine to me; maybe someone later on can look into whether its worth attempting to generalize the code for that case.

File: compiler/rustc_mir_transform/src/jump_threading.rs (hunk 20 above)
Comment (cjgillot) [reply compiler/rustc_mir_transform/src/jump_threading.rs:754]: This correction is a change in the definition. The restriction on loop headers is purely heuristic, and has no impact on soundness. LLVM does not like irreducible loops, so we avoid creating them.
The DFA test means precisely to capture this, but was overlooked. With a845bac6ba434483b483c0561011227c1eff9daf it should not break silently any more.

**Comment by cjgillot** [comment]
> Okay. That seems fine to me; maybe someone later on can look into whether its worth attempting to generalize the code for that case.

I'll think about it. I had an idea how to avoid the current backtracking algorithm, replaced by a single pass on the CFG. But there may be some time before I get around to write it.
